shm/src/ship.cpp

```diff
@@ -0,0 +1,99 @@
+#include "ship.hpp"
+
+#include <algorithm>
+#include <iostream>
+
+#include "fruit.hpp"
+
+Ship::Ship()
+    : id_(-1) {}
+Ship::Ship(size_t id, std::string name, size_t speed, size_t maxCrew, size_t capacity)
+    : id_(id), name_(name), speed_(speed), maxCrew_(maxCrew), capacity_(capacity) {}
+Ship::Ship(size_t id, std::string name, size_t speed, size_t maxCrew, size_t capacity, Delegate* delegate, std::shared_ptr<Time> time)
+    : id_(id), name_(name), speed_(speed), maxCrew_(maxCrew), capacity_(capacity), delegate_(delegate), time_(time) {
+        time_->addObserver(this);
+        crew_ = maxCrew_;
+    }
+Ship::Ship(size_t id, std::string name, size_t speed)
+    : Ship(id, name, speed, 0, 0) {}
+Ship::~Ship() {
+    time_->removeObserver(this);
+}
+
+void Ship::setName(const std::string& name) {
+    name_ = name;
+}
+
+void Ship::nextDay() {
+    for (auto el : getCargo()) {
+        //*el.get() += 1;
+        el.get()->nextDay();
```

Comment:
Wrong. Time should notify all observers. So register observer for Cargo. Instead, notify them when nextDay will be call for Ship.